File: dp/sumofsubsequence.py
import math, sys

#DSA05008
sys.stdin = open("E:/OneDrive - ptit.edu.vn/pro/dsa/input.txt", mode = 'r')
sys.stdout = open("E:/OneDrive - ptit.edu.vn/pro/dsa/output.txt", mode = 'w')

t = int(input())
for _ in range(t):
    n, s = map(int, input().split())
    a = list(map(int, input().split()))
    dp = [False] * 50000
    dp[0] = True
    for i in range(n):
        for j in range(s, a[i] - 1, -1):
            # dp[j] |= dp[j - a[i]]
            if dp[j - a[i]] == 1:
                dp[j] = True

    print('YES' if dp[s] else 'NO')

File: dp/sumofsubsequence.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

//idhere
#define forup(i, a, b)      for (int i = (a); i < (b); i++)
#define fordown(i, l, r)    for (int i = l; i >= r; i--)
#define print(a, x, y)      forup(i, x, y) cout << a[i] << " "
#define all(M)              M.begin(), M.end()
#define el                  "\n"

void fileio()
{
    #ifndef ONLINE_JUDGE
    freopen("E:/OneDrive - ptit.edu.vn/pro/dsa/input.txt", "r", stdin);
    freopen("E:/OneDrive - ptit.edu.vn/pro/dsa/output.txt", "w", stdout);
    #endif
}

int main()
{
    fileio();
    ios::sync_with_stdio(false); cin.tie(0);

    int t; cin >> t;
    while(t--)
    {
        int n, s; 
        cin >> n >> s;
        vector<int> a(n);
        forup(i, 0, n) cin >> a[i];
        vector<bool> dp(s + 1, 0);
        dp[0] = true;
        forup(i, 0, n)
        {
            fordown(j, s, a[i])
            {
                if (dp[j - a[i]])
                    dp[j] = 1;
            }
        }
        
        cout << (dp[s] ? "YES" : "NO") << el;
    }
}